File: AlphaStandardDemo/Abstracts/GenericViewModel.vb
Imports System.ComponentModel
Imports System.IO
Imports System.Windows.Input
Imports AlphaStandardDemo.Classes
Imports AlphaStandardDemo.Interfaces

Namespace Abstracts
  Public MustInherit Class GenericViewModel(Of TypeT As IEntity)
    Implements INotifyPropertyChanged

    Private ReadOnly Property _repository As GenericRepository(Of TypeT)

    Private _currentItem As TypeT
    Public Property CurrentItem() As TypeT
      Get
        Return _currentItem
      End Get
      Set(value As TypeT)
        _currentItem = value
        OnPropertyChanged(NameOf(CurrentItem))
      End Set
    End Property
    Private _listTypeT As BindingList(Of TypeT)
    Public Property ListTypeT() As BindingList(Of TypeT)
      Get
        Return _listTypeT
      End Get
      Set(value As BindingList(Of TypeT))
        _listTypeT = value
        OnPropertyChanged(NameOf(ListTypeT))
      End Set
    End Property
    Private _buttonNewEnabled As Boolean
    Public Property ButtonNewEnabled() As Boolean
      Get
        Return _buttonNewEnabled
      End Get
      Set(value As Boolean)
        _buttonNewEnabled = value
        OnPropertyChanged(NameOf(ButtonNewEnabled))
      End Set
    End Property
    Private _buttonUpdateEnabled As Boolean
    Public Property ButtonUpdateEnabled() As Boolean
      Get
        Return _buttonUpdateEnabled
      End Get
      Set(value As Boolean)
        _buttonUpdateEnabled = value
        OnPropertyChanged(NameOf(ButtonUpdateEnabled))
      End Set
    End Property
    Private _buttonDeleteEnabled As Boolean
    Public Property ButtonDeleteEnabled() As Boolean
      Get
        Return _buttonDeleteEnabled
      End Get
      Set(value As Boolean)
        _buttonDeleteEnabled = value
        OnPropertyChanged(NameOf(ButtonDeleteEnabled))
      End Set
    End Property
    Private _buttonCloseEnabled As Boolean
    Public Property ButtonCloseEnabled() As Boolean
      Get
        Return _buttonCloseEnabled
      End Get
      Set(value As Boolean)
        _buttonCloseEnabled = value
        OnPropertyChanged(NameOf(ButtonCloseEnabled))
      End Set
    End Property
    Private _buttonSaveEnabled As Boolean
    Public Property ButtonSaveEnabled() As Boolean
      Get
        Return _buttonSaveEnabled
      End Get
      Set(value As Boolean)
        _buttonSaveEnabled = value
        OnPropertyChanged(NameOf(ButtonSaveEnabled))
      End Set
    End Property
    Private _buttonCancelEnabled As Boolean
    Public Property ButtonCancelEnabled() As Boolean
      Get
        Return _buttonCancelEnabled
      End Get
      Set(value As Boolean)
        _buttonCancelEnabled = value
        OnPropertyChanged(NameOf(ButtonCancelEnabled))
      End Set
    End Property
    Private _controlsEnabled As Boolean
    Public Property ControlsEnabled() As Boolean
      Get
        Return _controlsEnabled
      End Get
      Set(value As Boolean)
        _controlsEnabled = value
        OnPropertyChanged(NameOf(ControlsEnabled))
      End Set
    End Property

    Public Sub New(repository As GenericRepository(Of TypeT))
      _repository = repository

      CurrentItem = Activator.CreateInstance(Of TypeT)
      ListTypeT = Activator.CreateInstance(Of BindingList(Of TypeT))
    End Sub

    Protected Overridable Sub Load()
      Try
        ReadMode()
        CurrentItem = Activator.CreateInstance(Of TypeT)
        ListTypeT = _repository.GetAll()
      Catch ex As Exception
        OnErrorOcurred(ex)
      End Try
    End Sub
    Protected Overridable Sub Save()
      Try
        If CurrentItem.Id = 0 Then
          _repository.Create(_currentItem)
        Else
          _repository.Update(_currentItem)
        End If

        _repository.SaveChanges()
        ReadMode()

        ListTypeT = _repository.GetAll()
        CurrentItem = ListTypeT.Last()
      Catch ex As Exception
        OnErrorOcurred(ex)
      End Try
    End Sub
    Protected Overridable Sub Create()
      Try
        EditMode()
        CurrentItem = Activator.CreateInstance(Of TypeT)
      Catch ex As Exception
        OnErrorOcurred(ex)
      End Try
    End Sub

    Protected Overridable Sub Cancel()
      Try
        ReadMode()
        If CurrentItem.Id = 0 Then CurrentItem = Activator.CreateInstance(Of TypeT)
      Catch ex As Exception
        OnErrorOcurred(ex)
      End Try
    End Sub
    Protected Overridable Sub Update()
      Try
        If CurrentItem Is Nothing OrElse CurrentItem.Id = 0 Then Throw New Exception("Selecione um registro")

        CurrentItem = _repository.GetById(_currentItem.Id)
        EditMode()
      Catch ex As Exception
        OnErrorOcurred(ex)
      End Try
    End Sub
    Protected Overridable Sub Delete()
      Try
        If CurrentItem.Id = 0 Then Throw New Exception("Selecione um registro")

        _repository.Delete(_currentItem.Id)
        _repository.SaveChanges()

        ListTypeT = _repository.GetAll()
        CurrentItem = Activator.CreateInstance(Of TypeT)
      Catch ex As Exception
        OnErrorOcurred(ex)
      End Try
    End Sub

    Public ReadOnly Property LoadCommand As ICommand
      Get
        Return New RelayCommand(AddressOf Load)
      End Get
    End Property
    Public ReadOnly Property SaveCommand As ICommand
      Get
        Return New RelayCommand(AddressOf Save)
      End Get
    End Property
    Public ReadOnly Property CreateCommand As ICommand
      Get
        Return New RelayCommand(AddressOf Create)
      End Get
    End Property
    Public ReadOnly Property CancelCommand As ICommand
      Get
        Return New RelayCommand(AddressOf Cancel)
      End Get
    End Property
    Public ReadOnly Property UpdateCommand As ICommand
      Get
        Return New RelayCommand(AddressOf Update)
      End Get
    End Property
    Public ReadOnly Property DeleteCommand As ICommand
      Get
        Return New RelayCommand(AddressOf Delete)
      End Get
    End Property

    Public Event ErrorOcurred As EventHandler(Of ErrorEventArgs)
    Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

    Protected Sub OnErrorOcurred(exception As Exception)
      RaiseEvent ErrorOcurred(Me, New ErrorEventArgs(exception))
    End Sub
    Protected Sub OnPropertyChanged(propertyName As String)
      RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
    End Sub

    Public Overridable Sub EditMode()
      ButtonNewEnabled = False
      ButtonUpdateEnabled = False
      ButtonDeleteEnabled = False
      ButtonCloseEnabled = False
      ButtonSaveEnabled = True
      ButtonCancelEnabled = True
      ControlsEnabled = True
    End Sub
    Public Overridable Sub ReadMode()
      ButtonNewEnabled = True
      ButtonUpdateEnabled = True
      ButtonDeleteEnabled = True
      ButtonCloseEnabled = True
      ButtonSaveEnabled = False
      ButtonCancelEnabled = False
      ControlsEnabled = False
    End Sub
  End Class
End Namespace

File: AlphaLib/Abstracts/GenericViewModel.cs
using AlphaLib.Classes;
using AlphaLib.Interfaces;
using System;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Windows.Input;

namespace AlphaLib.Abstracts
{
    public abstract class GenericViewModel<TypeT> : INotifyPropertyChanged where TypeT : class
    {
        private readonly GenericRepository<TypeT> _repository;

        private TypeT _currentItem;
        public TypeT CurrentItem
        {
            get => _currentItem;
            set
            {
                _currentItem = value;
                OnPropertyChanged(nameof(CurrentItem));
            }
        }

        private BindingList<TypeT> _listTypeT;
        public BindingList<TypeT> ListTypeT
        {
            get => _listTypeT;
            set
            {
                _listTypeT = value;
                OnPropertyChanged(nameof(ListTypeT));
            }
        }

        private bool _buttonNewEnabled;
        public bool ButtonNewEnabled
        {
            get => _buttonNewEnabled;
            set
            {
                _buttonNewEnabled = value;
                OnPropertyChanged(nameof(ButtonNewEnabled));
            }
        }

        private bool _buttonUpdateEnabled;
        public bool ButtonUpdateEnabled
        {
            get => _buttonUpdateEnabled;
            set
            {
                _buttonUpdateEnabled = value;
                OnPropertyChanged(nameof(ButtonUpdateEnabled));
            }
        }

        private bool _buttonDeleteEnabled;
        public bool ButtonDeleteEnabled
        {
            get => _buttonDeleteEnabled;
            set
            {
                _buttonDeleteEnabled = value;
                OnPropertyChanged(nameof(ButtonDeleteEnabled));
            }
        }

        private bool _buttonCloseEnabled;
        public bool ButtonCloseEnabled
        {
            get => _buttonCloseEnabled;
            set
            {
                _buttonCloseEnabled = value;
                OnPropertyChanged(nameof(ButtonCloseEnabled));
            }
        }

        private bool _buttonSaveEnabled;
        public bool ButtonSaveEnabled
        {
            get => _buttonSaveEnabled;
            set
            {
                _buttonSaveEnabled = value;
                OnPropertyChanged(nameof(ButtonSaveEnabled));
            }
        }

        private bool _buttonCancelEnabled;
        public bool ButtonCancelEnabled
        {
            get => _buttonCancelEnabled;
            set
            {
                _buttonCancelEnabled = value;
                OnPropertyChanged(nameof(ButtonCancelEnabled));
            }
        }

        private bool _controlsEnabled;
        public bool ControlsEnabled
        {
            get => _controlsEnabled;
            set
            {
                _controlsEnabled = value;
                OnPropertyChanged(nameof(ControlsEnabled));
            }
        }

        public GenericViewModel(GenericRepository<TypeT> repository)
        {
            _repository = repository;

            CurrentItem = Activator.CreateInstance<TypeT>();
            ListTypeT = new BindingList<TypeT>();
        }

        protected virtual void Load()
        {
            try
            {
                ReadMode();
                CurrentItem = Activator.CreateInstance<TypeT>();
                ListTypeT = _repository.GetAll();
            }
            catch (Exception ex)
            {
                OnErrorOcurred(ex);
            }
        }

        protected virtual void Save()
        {
            try
            {
                var item = (IEntity)CurrentItem;
                if (item.Id == 0)
                    _repository.Create(CurrentItem);
                else
                    _repository.Update(CurrentItem);

                _repository.SaveChanges();
                ReadMode();

                ListTypeT = _repository.GetAll();
                CurrentItem = ListTypeT.Last();
            }
            catch (Exception ex)
            {
                OnErrorOcurred(ex);
            }
        }

        protected virtual void Create()
        {
            try
            {
                EditMode();
                CurrentItem = Activator.CreateInstance<TypeT>();
            }
            catch (Exception ex)
            {
                OnErrorOcurred(ex);
            }
        }

        protected virtual void Cancel()
        {
            try
            {
                ReadMode();
                var item = (IEntity)CurrentItem;
                if (item.Id == 0)
                    CurrentItem = Activator.CreateInstance<TypeT>();
            }
            catch (Exception ex)
            {
                OnErrorOcurred(ex);
            }
        }

        protected virtual void Update()
        {
            try
            {
                var item = (IEntity)CurrentItem;
                if (CurrentItem == null || item.Id == 0)
                    throw new Exception("Selecione um registro");

                CurrentItem = _repository.GetById(item.Id);
                EditMode();
            }
            catch (Exception ex)
            {
                OnErrorOcurred(ex);
            }
        }

        protected virtual void Delete()
        {
            try
            {
                var item = (IEntity)CurrentItem;
                if (item.Id == 0)
                    throw new Exception("Selecione um registro");

                _repository.Delete(item.Id);
                _repository.SaveChanges();

                ListTypeT = _repository.GetAll();
                CurrentItem = Activator.CreateInstance<TypeT>();
            }
            catch (Exception ex)
            {
                OnErrorOcurred(ex);
            }
        }

        public ICommand LoadCommand => new RelayCommand(a => Load());
        public ICommand SaveCommand => new RelayCommand(a => Save());
        public ICommand CreateCommand => new RelayCommand(a => Create());
        public ICommand CancelCommand => new RelayCommand(a => Cancel());
        public ICommand UpdateCommand => new RelayCommand(a => Update());
        public ICommand DeleteCommand => new RelayCommand(a => Delete());

        public event EventHandler<ErrorEventArgs> ErrorOcurred;
        public event PropertyChangedEventHandler PropertyChanged;

        protected void OnErrorOcurred(Exception exception)
        {
            ErrorOcurred?.Invoke(this, new ErrorEventArgs(exception));
        }

        protected void OnPropertyChanged(string propertyName)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }

        public virtual void EditMode()
        {
            ButtonNewEnabled = false;
            ButtonUpdateEnabled = false;
            ButtonDeleteEnabled = false;
            ButtonCloseEnabled = false;
            ButtonSaveEnabled = true;
            ButtonCancelEnabled = true;
            ControlsEnabled = true;
        }

        public virtual void ReadMode()
        {
            ButtonNewEnabled = true;
            ButtonUpdateEnabled = true;
            ButtonDeleteEnabled = true;
            ButtonCloseEnabled = true;
            ButtonSaveEnabled = false;
            ButtonCancelEnabled = false;
            ControlsEnabled = false;
        }
    }
}